src/pool/srv_target.c

```diff
@@ -1377,42 +1377,72 @@ ds_pool_tgt_query_map_handler(crt_rpc_t *rpc)
 {
 	struct pool_tgt_query_map_in   *in = crt_req_get(rpc);
 	struct pool_tgt_query_map_out  *out = crt_reply_get(rpc);
-	struct ds_pool_hdl	       *hdl;
 	struct ds_pool		       *pool;
 	struct pool_buf		       *buf;
 	unsigned int			version;
 	int				rc;
 
-	D_DEBUG(DB_TRACE, DF_UUID": handling rpc %p\n",
-		DP_UUID(in->tmi_op.pi_uuid), rpc);
+	D_DEBUG(DB_TRACE, DF_UUID": handling rpc %p: hdl="DF_UUID"\n",
+		DP_UUID(in->tmi_op.pi_uuid), rpc, DP_UUID(in->tmi_op.pi_hdl));
 
-	hdl = ds_pool_hdl_lookup(in->tmi_op.pi_hdl);
-	if (hdl == NULL) {
-		rc = -DER_NO_HDL;
-		goto out;
+	/* Validate the pool handle and get the ds_pool object. */
+	if (daos_rpc_from_client(rpc)) {
+		struct ds_pool_hdl *hdl;
+
+		hdl = ds_pool_hdl_lookup(in->tmi_op.pi_hdl);
+		if (hdl == NULL) {
+			rc = -DER_NO_HDL;
+			goto out;
+		}
+		ds_pool_get(hdl->sph_pool);
+		pool = hdl->sph_pool;
+		ds_pool_hdl_put(hdl);
+	} else {
+		/*
+		 * See the comment on validating the pool handle in
+		 * ds_pool_query_handler.
+		 */
+		pool = ds_pool_lookup(in->tmi_op.pi_uuid);
+		if (pool == NULL) {
+			D_ERROR(DF_UUID": failed to look up pool\n", DP_UUID(in->tmi_op.pi_uuid));
+			rc = -DER_NONEXIST;
+			goto out;
+		}
+		rc = ds_pool_hdl_is_from_srv(pool, in->tmi_op.pi_hdl);
```

Comment:
you can try to use sp_srv_pool_hdl in ds_pool first? instead of issuing IV here?

Reply:
@wangdi1, do you mean something like the following? I struggle to understand whether pool->sp_srv_pool_hdl can be stale (i.e., generated by a previous PS leader and different from the one generated by the current PS leader).

    --- a/src/pool/srv_pool.c
    +++ b/src/pool/srv_pool.c
    @@ -5716,6 +5716,9 @@ ds_pool_hdl_is_from_srv(struct ds_pool *pool, uuid_t hdl)
            uuid_t  srv_hdl;
            int     rc;
     
    +       if (!uuid_is_null(pool->sp_srv_pool_hdl))
    +               return uuid_compare(pool->sp_srv_pool_hdl, hdl) == 0;
    +
            rc = ds_pool_iv_srv_hdl_fetch(pool, &srv_hdl, NULL);
            if (rc != 0)
                    return rc;

Reply:
@wangdi1, any thought on why sp_srv_pool_hdl can never be stale?

Reply:
@liw sp_srv_pool_hdl is generated by PS leader during step up process, then broadcast to all servers  by IV.  For stable, you mean consistent cross the PS leader change? This seems unlikely, unless we put it to RDB.

Reply:
oh, I mean

if (uuid_compare(pool->sp_srv_pool_hdl, hdl) != 0) {
          rc = ds_pool_hdl_is_from_srv(pool, pi_hdl);

          /* extra check */
}

Or even move this uuid_compare(pool->sp_srv_pool_hdl, hdl) inside ds_pool_hdl_is_from_srv(),  so it does not have to create extra ULT each time?